src/renderers/WebGLRenderer.js

```diff
@@ -1340,6 +1340,8 @@ function WebGLRenderer( parameters = {} ) {
 		object.modelViewMatrix.multiplyMatrices( camera.matrixWorldInverse, object.matrixWorld );
 		object.normalMatrix.getNormalMatrix( object.modelViewMatrix );
 
+		material.onUpdate( _this, scene, camera, geometry, object, group );
```

Comment:
> For this done correctly, is needed a frame update for the nodes, this update can be done in with Material.onUpdate()

If the material is used for more than one 3D object, this callback will be executed twice. Would this be okay? Otherwise we need a mechanism similar to `WebGLObjects`:

https://github.com/mrdoob/three.js/blob/fb43e1fe4f4b484131d2f130f5b7410cabb328ed/src/renderers/webgl/WebGLObjects.js#L12-L20

Reply:
I commit my approach for this I was waiting make this in a next PR because I have doubts if I can improve this... But as you realized this very well: https://github.com/mrdoob/three.js/pull/22417/commits/d511c02c6d43a22c0636acce74d359ba6f1ca62b

Reply:
That is a good workaround. I just wonder if we actually need `Object3D.onUpdate()` or something like that?

Reply:
This already happens here: https://github.com/mrdoob/three.js/blob/d511c02c6d43a22c0636acce74d359ba6f1ca62b/examples/jsm/renderers/webgl/nodes/WebGLNodes.js#L17

Material.onUpdate() is a optimized way to update the nodes before the object has rendered. Otherside can transverse the scene and update nodes before of the `.render()` but I think that we would lose some feature like `webgpu_instance_uniform`

Reply:
Yeah, the WebGL and WebGPU renderer are different in that regard. The WebGPU implementation is actually more correct since it maintains a builder per render item. In WebGL we have to use the material as the key for and circumvent the respective consequences. However, this is not a node material issue. It depends on the fact that `WebGLRenderer` maintains multiple shader programs per material.

 So the current approach is okay an `onUpdate()` seems a good solution!